asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go

```diff
@@ -260,63 +242,70 @@ func (t *SimpleChaincode) GetAssetsByRange(ctx contractapi.TransactionContextInt
+	// Execute a key range query on all keys starting with 'color'
 	coloredAssetResultsIterator, err := ctx.GetStub().GetStateByPartialCompositeKey("color~name", []string{color})
 	if err != nil {
-		return fmt.Errorf(err.Error())
+		return err
 	}
 	defer coloredAssetResultsIterator.Close()
 
-	// Iterate through result set and for each asset found, transfer to newOwner
-	var i int
-	for i = 0; coloredAssetResultsIterator.HasNext(); i++ {
-		// Note that we don't get the value (2nd return variable), we'll just get the asset name from the composite key
+	assets := map[string][]byte{}
+	for coloredAssetResultsIterator.HasNext() {
 		responseRange, err := coloredAssetResultsIterator.Next()
 		if err != nil {
-			return fmt.Errorf(err.Error())
+			return err
 		}
 
-		// get the color and name from color~name composite key
 		_, compositeKeyParts, err := ctx.GetStub().SplitCompositeKey(responseRange.Key)
 		if err != nil {
-			return fmt.Errorf(err.Error())
+			return err
 		}
 
+		// To provide as much determinism as we can on this bulk transfer operation we first validate
+		// that each asset exists, and can be marshalled before doing a bulk PutState.
+		// This will, to the best of our ability, make sure later PutStates don't fail after
+		// early PutStates succeed. The assets are stored in a map after they have been validated
+		// and then a bulk PutState is performed in a separate operation.
 		if len(compositeKeyParts) > 2 {
 			returnedAssetID := compositeKeyParts[1]
-
-			// Now call the transfer function for the found asset.
-			// Re-use the same function that is used to transfer individual assets
-			err = t.TransferAsset(ctx, returnedAssetID, newOwner)
-			// if the transfer failed break out of loop and return error
+			asset, err := t.ReadAsset(ctx, returnedAssetID)
+			if err != nil {
+				return err
+			}
+			asset.Owner = strings.ToLower(newOwner)
+			assetBytes, err := json.Marshal(asset)
 			if err != nil {
-				return fmt.Errorf("Transfer failed: %v", err)
+				return err
 			}
+			assets[returnedAssetID] = assetBytes
+		}
+	}
+
+	for id, asset := range assets {
+		err = ctx.GetStub().PutState(id, asset)
+		if err != nil {
+			return fmt.Errorf("transfer failed for asset %s: %v", id, err)
 		}
 	}
 
 	return nil
 }
 
-// QueryAssetsByOwner queries for assets based on a passed in owner.
+// QueryAssetsByOwner queries for assets based on the owners name.
 // This is an example of a parameterized query where the query logic is baked into the chaincode,
 // and accepting a single query parameter (owner).
 // Only available on state databases that support rich query (e.g. CouchDB)
 // Example: Parameterized rich query
 func (t *SimpleChaincode) QueryAssetsByOwner(ctx contractapi.TransactionContextInterface, owner string) ([]*QueryResult, error) {
-	queryString := fmt.Sprintf("{\"selector\":{\"docType\":\"asset\",\"owner\":\"%s\"}}", owner)
-
+	queryString := fmt.Sprintf(`{"selector":{"docType":"asset","owner":"%s"}}`, owner)
```

Comment:
owner `ToLower` in the query

couchdb index is `"type":"json"` , so case sensitive

Reply:
I've got to find the comment from another PR earlier today, but all calls `.ToLower()` should be removed. That was an artifact of the old chaincode where users didn't provide the owners name directly (it was generated for them). Now that the users can provide the owners name, case should be up to them entirely